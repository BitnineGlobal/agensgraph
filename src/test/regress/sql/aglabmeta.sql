--
-- AgensGraph catalog aglabmeta test
--
CREATE GRAPH labmeta;
SET graph_path = labmeta;

-- create edge

CREATE (:human)-[:know]->(:human {age:1});
CREATE (:human)-[:know]->(:human {age:2});
CREATE (:human)-[:know]->(:human {age:3});
CREATE (:dog)-[:follow]->(:human);
CREATE (:dog)-[:likes]->(:dog);

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- create multiple edges

CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);
CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);
CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- create repeated edges;

CREATE (:human)-[:know]->(:human)-[:know]->(:human)-[:know]->(:human)-[:know]->(:human);

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- delete edge

MATCH (a)-[r:love]->(b)
DELETE r;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- drop elabel

DROP ELABEL hate CASCADE;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- drop vlabel

DROP VLABEL human CASCADE;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- drop graph

DROP GRAPH labmeta CASCADE;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- Sub Transaction

-- Rollback
CREATE GRAPH labmeta;

BEGIN;
	CREATE (:dog)-[:follow]->(:human);
	SAVEPOINT sv1;
	CREATE (:dog)-[:likes]->(:cat);
	ROLLBACK TO SAVEPOINT sv1;
	CREATE (:human)-[:love]->(:dog);
	SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
COMMIT;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- Release
BEGIN;
	CREATE (:dog)-[:likes]->(:cat);
	SAVEPOINT sv2;
	CREATE (:cat)-[:ignore]->(:dog);
	RELEASE SAVEPOINT sv2;
COMMIT;

SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";

-- cleanup

DROP GRAPH labmeta CASCADE;
