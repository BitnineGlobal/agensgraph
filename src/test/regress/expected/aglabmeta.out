--
-- AgensGraph catalog aglabmeta test
--
CREATE GRAPH labmeta;
SET graph_path = labmeta;
-- create edge
CREATE (:human)-[:know]->(:human {age:1});
CREATE (:human)-[:know]->(:human {age:2});
CREATE (:human)-[:know]->(:human {age:3});
CREATE (:dog)-[:follow]->(:human);
CREATE (:dog)-[:likes]->(:dog);
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | dog   |         1
 labmeta   | human | know   | human |         3
(3 rows)

-- create multiple edges
CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);
CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);
CREATE (:human)-[:know]->(:human)-[:follow]->(:human)-[:hate]->(:human)-[:love]->(:human);
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | dog   |         1
 labmeta   | human | follow | human |         3
 labmeta   | human | hate   | human |         3
 labmeta   | human | know   | human |         6
 labmeta   | human | love   | human |         3
(6 rows)

-- create repeated edges;
CREATE (:human)-[:know]->(:human)-[:know]->(:human)-[:know]->(:human)-[:know]->(:human);
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | dog   |         1
 labmeta   | human | follow | human |         3
 labmeta   | human | hate   | human |         3
 labmeta   | human | know   | human |        10
 labmeta   | human | love   | human |         3
(6 rows)

-- delete edge
MATCH (a)-[r:love]->(b)
DELETE r;
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | dog   |         1
 labmeta   | human | follow | human |         3
 labmeta   | human | hate   | human |         3
 labmeta   | human | know   | human |        10
(5 rows)

-- drop elabel
DROP ELABEL hate CASCADE;
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | dog   |         1
 labmeta   | human | follow | human |         3
 labmeta   | human | know   | human |        10
(4 rows)

-- drop vlabel
DROP VLABEL human CASCADE;
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start | edge  | end | edgecount 
-----------+-------+-------+-----+-----------
 labmeta   | dog   | likes | dog |         1
(1 row)

-- drop graph
DROP GRAPH labmeta CASCADE;
NOTICE:  drop cascades to 8 other objects
DETAIL:  drop cascades to sequence labmeta.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to elabel know
drop cascades to vlabel dog
drop cascades to elabel follow
drop cascades to elabel likes
drop cascades to elabel love
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start | edge | end | edgecount 
-----------+-------+------+-----+-----------
(0 rows)

-- Sub Transaction
-- Rollback
CREATE GRAPH labmeta;
BEGIN;
	CREATE (:dog)-[:follow]->(:human);
	SAVEPOINT sv1;
	CREATE (:dog)-[:likes]->(:cat);
	ROLLBACK TO SAVEPOINT sv1;
	CREATE (:human)-[:love]->(:dog);
	SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start | edge | end | edgecount 
-----------+-------+------+-----+-----------
(0 rows)

COMMIT;
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | dog   | follow | human |         1
 labmeta   | human | love   | dog   |         1
(2 rows)

-- Release
BEGIN;
	CREATE (:dog)-[:likes]->(:cat);
	SAVEPOINT sv2;
	CREATE (:cat)-[:ignore]->(:dog);
	RELEASE SAVEPOINT sv2;
COMMIT;
SELECT * FROM ag_labname_meta ORDER BY start, edge, "end";
 graphname | start |  edge  |  end  | edgecount 
-----------+-------+--------+-------+-----------
 labmeta   | cat   | ignore | dog   |         1
 labmeta   | dog   | follow | human |         1
 labmeta   | dog   | likes  | cat   |         1
 labmeta   | human | love   | dog   |         1
(4 rows)

-- cleanup
DROP GRAPH labmeta CASCADE;
NOTICE:  drop cascades to 10 other objects
DETAIL:  drop cascades to sequence labmeta.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel dog
drop cascades to elabel follow
drop cascades to vlabel human
drop cascades to elabel love
drop cascades to elabel likes
drop cascades to vlabel cat
drop cascades to elabel ignore
